create instances via nodegroup

gthao313 opened the pull request:
<!--
Tips:
- Please read the CONTRIBUTING document to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
#162


**Description of changes:**
Currently we create bottlerocket nodes via launching instances directly, and in this PR we're going to create bottlerocket node via nodegroups. From this we can specify any arbitrary setting we want. and we can have the node-labels set as a Bottlerocket settings directly.

Tasks

- Nodegroup creation and termination
- Launch template creation and termination
- One-time iam role and profile creation and deletion
- Mapping iam role and profile to cluster


**Testing done:**
Test on ipv4 and ipv6 clusters
```
cargo run --bin integ integration-test --cluster-name bottlerocket-test --region us-west-2 --bottlerocket-version 1.5.0  --arch x86_64 --nodegroup-name=tianhg-test-x64
```

```
cargo run --bin integ clean --cluster-name bottlerocket-test --region us-west-2 --nodegroup-name=tianhg-test-x64
```


```
cargo run --bin integ integration-test --cluster-name bottlerocket-test --region us-west-2 --bottlerocket-version 1.5.0  --arch arm64 --nodegroup-name=tianhg-test-arm64
```
```
cargo run --bin integ clean --cluster-name bottlerocket-test --region us-west-2  --nodegroup-name=tianhg-test-arm64
```



**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

gthao313 committed:
create instances via nodegroup

cbgbt commented on integ/src/nodegroup_provider.rs:
```diff
@@ -0,0 +1,573 @@
+    Ok(())
+}
+
+fn eks_role_policy_document() -> String {
```
Should this just be a `const &str`?

cbgbt commented on integ/src/nodegroup_provider.rs:
```diff
@@ -0,0 +1,573 @@
+        .await;
+
+    match nodegroup {
+        Err(_resource_not_found_exception) => Ok(true),
+        _ => Ok(false),
```
Are these bools reversed? I would expect `resource_not_found` to result in `false`.

gthao313 replied:
I think the function name `confirm_nodegroup_exists` is not accurate, and I'll rename it to `confirm_nodegroup_deleted`. thanks

cbgbt commented on integ/src/nodegroup_provider.rs:
```diff
@@ -0,0 +1,573 @@
+        .build()
+}
+
+fn userdata(endpoint: &str, cluster_name: &str, certificate: &str) -> String {
```
We should probably use a [TOML serializer](https://docs.rs/toml/latest/toml/ser/fn.to_string_pretty.html) with serde for this, since it would be pretty easy to break the TOML formatting with the right input.

gthao313 replied:
Do you mean something like below? I tried this but `create_launch_template().user_data( input)`, the input need a valid base64 string. Can I still use TOML serializer to output base64 string? 
```
n userdata(endpoint: &str, cluster_name: &str, certificate: &str) -> ProviderResult<String> {
    let user_data = toml::ser::to_string_pretty(&format!(
        r#"[settings.updates]
            ignore-waves = true
                
            [settings.kubernetes]
            api-server = "{}"
            cluster-name = "{}"
            cluster-certificate = "{}"
        "#,
        endpoint, cluster_name, certificate
    ))
    .context("Unable convert userdata to pretty string")?;

    Ok(user_data)
```

cbgbt replied:
I meant more like create a series of structs like:

```rust
#[derive(Serialize)]
struct Settings {
    updates: UpdateSettings,
    kubernetes: KubernetesSettings,
}

#[derive(Serialize)]
struct UpdateSettings {
    ignore_waves: bool,
}

#[derive(Serialize)]
struct KubernetesSettings {
    api_server: String,
    // etc...
}
```

Then you could do something like:

```rust
let settings = Settings {
    updates: UpdateSettings {
        ignore_waves: true,
    },
    kubernetes: KubernetesSettings {
        // etc...
    },
}
toml::ser::to_string_pretty(settings);
```

gthao313 replied:
I tried the method that @cbgbt mentioned above; however, the string formate of userdata which was created by `toml::ser::to_string_pretty()` can be recognized by 'early-boot-config'. We can't use this method and I think it's better to keep the original method.
Error
```
[    5.826817] early-boot-config[1483]: Provider error: Unable to serialize settings from instance user data: TOML data did not contain 'settings' section
```
userdata created by toml
```
"[\"settings.updates\"]\n
ignore-waves = true\n\n
[\"settings.kubernetes\"]\n
api-server = 'xxxx7.us-west-2.eks.amazonaws.com'\n
cluster-name = 'bottlerocket-test'\n
cluster-certificate = '=xxxxx'\n"
```

gthao313 commented on integ/src/pods-template.yaml:
```diff
@@ -31,7 +31,7 @@ spec:
       containers:
       - name: nginx
-        image: k8s.gcr.io/nginx-slim:0.8
+        image: nginx:1.21.6
```
Update nginx image version to `1.21.6` so that able to create arm/amd pods. Old version Nginx doesn't support arm.

arnaldo2792 replied:
Should you have a different commit for this change?

gthao313 commented:
Push above address comments, update third party packages.

gthao313 commented on integ/src/nodegroup_provider.rs:
```diff
@@ -0,0 +1,623 @@
+        .build()
+}
+
+fn userdata(
```
resolve conflict on this part. The code is from [IPV6 PR](https://github.com/bottlerocket-os/bottlerocket-update-operator/commit/e803d0eaa2f4be468415ea047627df33201b832a)

gthao313 commented:
Push above rebase and update aws rust dependencies to `0.13.0`

arnaldo2792 commented on integ/src/main.rs:
```diff
@@ -27,11 +28,11 @@ const DEFAULT_KUBECONFIG_FILE_NAME: &str = "kubeconfig.yaml";
 
-/// This value configure how long it sleeps between create instance and label instance.
-const INTEGRATION_TEST_DELAY: Duration = Duration::from_secs(60);
+// The default name for nodegroup
```
pedantic nit:

```suggestion
// The default name for the nodegroup
```

arnaldo2792 commented on integ/src/main.rs:
```diff
@@ -232,38 +223,55 @@ async fn run() -> Result<()> {
+            if !nodes_exist(k8s_client).await.context(error::RunBrupop)? {
+                // Clean up all brupop resources like namespace, deployment on brupop test
+                info!(
+                    "Deleting all brupop cluster resources which created by integration test ..."
```
nit:

```suggestion
                    "Deleting all brupop cluster resources created by integration test ..."
```

arnaldo2792 commented on integ/src/updater.rs:
```diff
@@ -195,6 +148,18 @@ pub async fn process_pods_test(action: Action, kube_config_path: &str) -> Update
+// Find if any node is running in cluster
+pub async fn nodes_exist(k8s_client: kube::client::Client) -> UpdaterResult<bool> {
+    let nodes: Api<Node> = Api::all(k8s_client.clone());
+    // let a = nodes.list_node().await.context(update_error::FindBrupopPods {})?;
```
Extra comment here?

arnaldo2792 commented on integ/src/main.rs:
```diff
@@ -232,38 +223,55 @@ async fn run() -> Result<()> {
 
-            // delete all created pods for testing.
-            info!(
+            // Terminate nodegroup which created by integration test.
```
pendantic nit:

```suggestion
            // Terminate nodegroup created by integration test.
```

arnaldo2792 commented on integ/src/updater.rs:
```diff
@@ -195,6 +148,18 @@ pub async fn process_pods_test(action: Action, kube_config_path: &str) -> Update
     Ok(())
 }
 
+// Find if any node is running in cluster
```
nit:

```suggestion
// Find if any node is running in the cluster
```

gthao313 committed:
update nginx

Update nginx image version to 1.21.6 so that able to create arm/amd pods.
Old version Nginx doesn't support arm.
